clang/lib/Sema/SemaChecking.cpp

```diff
@@ -5553,6 +5555,112 @@ bool Sema::BuiltinSetjmp(CallExpr *TheCall) {
+      UO && UO->getOpcode() == UO_AddrOf) {
+    Arg = UO->getSubExpr()->IgnoreParenImpCasts();
+
+    if (auto *ASE = dyn_cast<ArraySubscriptExpr>(Arg))
+      Arg = ASE->getBase()->IgnoreParenImpCasts();
+  }
+
+  if (const MemberExpr *ME = dyn_cast_if_present<MemberExpr>(Arg)) {
+    if (!ME->isFlexibleArrayMemberLike(
+            Context, getLangOpts().getStrictFlexArraysLevel()))
+      return Diag(Arg->getBeginLoc(),
+                  diag::err_builtin_counted_by_ref_must_be_flex_array_member)
+             << Arg->getSourceRange();
+
+    if (ME->getMemberDecl()->getType()->isCountAttributedType()) {
+      if (FieldDecl *FAMDecl = dyn_cast<FieldDecl>(ME->getMemberDecl())) {
+        if (FieldDecl *CountFD = FAMDecl->findCountedByField()) {
+          // Reverse through any anonymous structs / unions surrounding the
+          // flexible array member. We'll build any necessary MemberExpr's to
+          // anonymous structs / unions when building a reference to the
+          // 'count' field.
+          RecordDecl *RD = FAMDecl->getParent();
+          DeclContext *DC = RD;
+          for (; DC->isRecord(); DC = DC->getLexicalParent()) {
+            if (!RD->isAnonymousStructOrUnion())
+              break;
+            RD = cast<RecordDecl>(DC);
+            if (auto *Base = dyn_cast<MemberExpr>(ME->getBase()))
+              ME = Base;
+          }
+
+          // See if the count's FieldDecl is within anonymous structs.
+          SmallVector<NamedDecl *, 2> PathToFD;
+          for (Decl *D : RD->decls()) {
+            if (auto *IFD = dyn_cast<IndirectFieldDecl>(D);
+                IFD && IFD->getAnonField() == CountFD) {
+              PathToFD.insert(PathToFD.begin(), IFD->chain_begin(),
+                              IFD->chain_end());
+              break;
+            }
+          }
+
+          if (PathToFD.empty())
+            PathToFD.push_back(CountFD);
+
+          // Build a MemberExpr to the 'count' field. This accounts for any
+          // anonymous structs / unions that may contain the field.
+          bool isArrow = ME->isArrow();
+          Expr *New = ME->getBase();
+          for (NamedDecl *ND : PathToFD) {
+            ValueDecl *VD = cast<ValueDecl>(ND);
+            New = MemberExpr::CreateImplicit(Context, New, isArrow, VD,
+                                             VD->getType(), VK_PRValue,
+                                             OK_Ordinary);
+            isArrow = false;
+          }
+
+          QualType CountTy = CountFD->getType();
+          New = UnaryOperator::Create(Context, New, UO_AddrOf,
+                                      Context.getPointerType(CountTy),
+                                      VK_LValue, OK_Ordinary, SourceLocation(),
+                                      false, FPOptionsOverride());
+
+          return ExprResult(New);
+        } else {
+          llvm::report_fatal_error("Cannot find the counted_by 'count' field");
```

Comment:
Why is this using `report_fatal_error` rather than either issuing a regular diagnostic, asserting that the field is always found, or otherwise recovering?

Reply:
At this point it should be there, but I changed it into a regular diagnostic.